internal/cmd/gocdk/launch.go

```diff
@@ -15,11 +15,170 @@
+		// TODO(light): Pass environment variables from tfOutput into env.
+	})
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+	fmt.Fprintf(pctx.stdout, "Serving at %s\n", launchURL)
+	return nil
+}
+
+// Launcher is the interface for any type that can launch a Docker image.
+type Launcher interface {
+	Launch(ctx context.Context, input *launchInput) (*url.URL, error)
+}
+
+// launchInput is the input to a launcher.
+type launchInput struct {
+	// dockerImage specifies the image name and tag of the local Docker image to
+	// deploy. If the local image does not exist, then the launcher should return
+	// an error.
+	dockerImage string
+
+	// env is the set of additional environment variables to set. It should not
+	// include PORT nor should it contain multiple entries for the same variable
+	// name.
+	env []string
+
+	// specifier is the set of arguments passed from a biome's Terraform module.
+	specifier launchSpecifier
+}
+
+// newLauncher creates the launcher for the given name.
+func newLauncher(ctx context.Context, pctx *processContext, launcherName string) (Launcher, error) {
+	logger := log.New(pctx.stderr, "gocdk: ", log.Ldate|log.Ltime)
+	switch launcherName {
+	case "local":
+		return &localLauncher{
+			logger:    logger,
+			dockerEnv: pctx.env,
+			dockerDir: pctx.workdir,
+		}, nil
+	default:
+		return nil, xerrors.Errorf("prepare launcher: unknown launcher %q", launcherName)
+	}
+}
+
+// localLauncher starts local Docker containers.
+type localLauncher struct {
+	logger    *log.Logger
+	dockerEnv []string
+	dockerDir string
+}
+
+// Launch implements Launcher.Launch.
+func (local *localLauncher) Launch(ctx context.Context, input *launchInput) (*url.URL, error) {
+	dockerArgs := []string{
+		"run",
+		"--rm",
+		"--detach",
+		"--publish", "8080:8080",
+	}
+	for _, v := range input.env {
+		dockerArgs = append(dockerArgs, "--env", v)
+	}
+	dockerArgs = append(dockerArgs, "--env", "PORT=8080")
+	dockerArgs = append(dockerArgs, input.dockerImage)
+
+	c := exec.CommandContext(ctx, "docker", dockerArgs...)
+	c.Env = local.dockerEnv
+	c.Dir = local.dockerDir
+	out, err := c.CombinedOutput()
+	if err != nil {
+		if len(out) == 0 {
+			return nil, xerrors.Errorf("local launch: docker run: %w", err)
+		}
+		return nil, xerrors.Errorf("local launch: docker run:\n%s", out)
+	}
+
+	local.logger.Printf("Docker container %s started, waiting for healthy...", bytes.TrimSuffix(out, []byte("\n")))
+	serveURL := &url.URL{
+		Scheme: "http",
+		Host:   "localhost:8080",
+		Path:   "/",
+	}
+	healthCheckURL := &url.URL{
+		Scheme: serveURL.Scheme,
+		Host:   serveURL.Host,
+		Path:   "/healthz/readiness",
+	}
+	if err := waitForHealthy(ctx, healthCheckURL); err != nil {
+		return nil, xerrors.Errorf("local launch: %w", err)
+	}
+	return serveURL, nil
+}
+
+// launchSpecifier is the set of arguments passed from a biome's Terraform
+// module to the biome's launcher.
+type launchSpecifier map[string]interface{}
```

Comment:
I am not sure that this type is adding much value. Consider leaving it as a `map[string]interface{}`. You can have still have a `stringValue` helper that takes one as an argument.

Reply:
Sorry, I don't follow your comment. The extra type made this harder to follow for me, not easier. We'll probably be reading config variables from the json config in various places, I think it will be easier to follow if we just have a set of utilities for doing so rather than adding a new type for each use case.

Reply:
Do you mind removing it for now? If you feel like it needs to be added to make the Cloud Run launcher work well, you can reintroduce it in that PR, but at this point I think it just adds a level of indirection that's hard to follow and doesn't provide much value.

Reply:
Removed.